### PR TITLE
added debounce functionality to the ModalSelectorLayout component

### PR DESCRIPTION
**Pull request by bt4R9**
### 🔦 Summary
Just added a debounce functionality to the Search component inside a ModalSelectorLayout.

### ✅ Checklist
- 📚 Change is documented
  - [x] Story
  - [x] API description
- 🔬 Change is tested
  - [x] Component
  - [x] Visual test

### CODE REVIEW

File: src/ModalSelectorLayout/docs/index.story.js [hunk 1]
```diff
@@ -32,6 +32,7 @@ export default {
     itemsPerPage: 4,
     imageSize: 'large',
     withSearch: true,
+    searchDebounceMS: 100,
```
Comment (argshook) [inline src/ModalSelectorLayout/docs/index.story.js]:
```suggestion
    searchDebounceMs: 100,
```

**Comment by bt4R9** [comment]
Hi! I'm not sure we should change the behaviour of the `Search` component since it's controlled by `ModalSelectorLayout`. I mean we should provide the value immediately to the `Search` component in order to update the input value, but at the same time we want to get debounced onChange callback. Providing `searchDebounceMs` to the `Search` literally means that the component is no longer controlled by the parent component but we also want an onChange callback that makes current API of the `Search` inconsistent. What do you thing?

**Comment by bt4R9** [comment]
@argshook So in the last commit I moved debounce from the `ModalSelectorLayout` to the `Search` but the API I came up with is kinda inconsistent now because `_isControlled` is in conflict with `_isDebounced` which makes the API ugly. May be you have an idea how to save the old behaviour and smoothly implement debounce functionality? thx

File: src/Search/Search.js [hunk 2]
```diff
@@ -57,19 +79,25 @@ export default class Search extends WixComponent {
...
       this.props.onChange(e);
     } else {
       this.setState({
         inputValue: e.target.value,
       });
     }
+
+    if (this._isDebounced) {
+      // need a copy of the event because of the dobunce
+      this._onChangeDebounced({ ...e });
```
Comment (bt4R9) [inline src/Search/Search.js]: could be `e.persist()`, have no idea what approach is better :)

**Comment by argshook** [comment]
@bt4R9 sorry for taking a while. I submitted a PR to your branch with some changes. please see if it covers your cases and lets merge the functionality 
https://github.com/bt4R9/wix-style-react/pull/1

**Comment by bt4R9** [comment]
@argshook Hi, thx! The only thing I'd like change is the name of `_makeDebounced` method since internally it can both debounced and not debounced. How about `this._onChange = this._makeOnChange();`?

**Comment by argshook** [comment]
> @argshook Hi, thx! The only thing I'd like to change is the name of `_makeDebounced` method since internally it can be both debounced and not debounced. How about `this._onChange = this._makeOnChange();`?

sure thing, you can push changes if you have more in mind :)

let me know when it's ready to merge

**Comment by bt4R9** [comment]
@argshook Let's merge it :)

**Comment by argshook** [comment]
@bt4R9 oh, i just noticed that only ModalSelectorLayout has tests for debounce functionality, but not Search component. That's no good, because ModalSelectorLayout could change or get removed and so tests together with it.

would you be willing to add tests for Search too?
i can do that also but am busy as usual, it'll take time till i reach it

**Comment by bt4R9** [comment]
Sure.

**Comment by bt4R9** [comment]
check it plz
may be we need more cases to check?

**Comment by argshook** [comment]
> may be we need more cases to check?

unless you can think of more. i mostly want tests for regression, so that 6 months from now they would cover us

if build is green, i'll merge, thanks for your work!

**Comment by bt4R9** [comment]
I'm not sure why the build is red. It seems like we have flaking e2e tests =/

**Comment by argshook** [comment]
yeah, we do have visual tests that like to timeout occasionally.
It's ok now, thanks for your work!

**Pull request merged**